### DISCUSSION TITLE
DownloadStream is here !

### DISCUSSION THREAD

**Discussion by Touffy** [Touffy/client-zip#66] [Announcements]
and it will soon be mentioned in the README when I have ran a longer series of tests, but you heard it here first…

[dl-stream](https://github.com/Touffy/dl-stream) is a very small library that exports the `DownloadStream` constructor. It has one job : read URLs and Requests from its source, and `fetch` them to give Responses for its consumer. You can see why that's relevant to client-zip, but just how well does DownloadStream do its job ?

Well, you remember those recent performance tests I published in the README, where fast libraries like client-zip and fflate take over two seconds to bundle 6214 small files downloaded by my classic async generator in Safari, and Conflux takes a bit less using a simple stream ? Just by pulling the files from a DownloadStream instead, client-zip and fflate take about one second each (and Conflux is down to 750ms). That's a **100% performance improvement** without touching the ZIP library itself.

Naturally, you won't see that much improvement with larger files. The whole point is to avoid making client-zip wait for the next file once it's finished with the previous one.

**Comment by artyuum** [comment]
I believe it's not possible to predict the zip size anymore when using dl-stream?

**Reply by Touffy** [reply]
It is, but you need to first get the metadata (without fetching the actual files) and pass the metadata to downloadZip. Ideally, you would use a dedicated route to provide the metadata for many files efficiently with a single request, instead of fetching all the files one by one and reading the headers.

**Comment by spookyuser** [comment]
Thanks sm for this and client-zip - they're amazing, just wondering, if I wanted to use dl-stream with a specific filename per string/request like {url, filename} is that possible? I was thinking I might be able to do it with a transform response like in the conflux example, but I can't seem to figure out exactly how to do that. Thank you!

**Reply by Touffy** [reply]
Of course it's possible. *dl-stream* returns a stream of Responses, so you need to transform each Response into a custom input object for client-zip (i.e. `{input, name}`) to specify a different filename than what client-zip would guess from the Response alone.

The *input* will be the Response itself, and *name* is up to you. Either you can compute the filename from the Response's properties, or you may use a predefined Map to lookup the Response's URL (for example) and get a name. For example :
```js
// keys are URLs, values are the desired filename for each
const files = new Map([
  ["https://example.org/file1.csv", "customName1.csv"],
  ["https://example.org/file2.xml", "customName2.xml"],
])
const responses = new DownloadStream(files.keys())
async function *inputForZip() {
  for await (const response of responses)
    yield { input: response, name: files.get(response.url) }
}
const blob = await downloadZip(inputForZip).blob()
```

**Reply by spookyuser** [reply]
Thank you so much!! That helped me a lot, could i make a PR with this as an example to the dl-stream readme?

**Reply by Touffy** [reply]
This is too specific to client-zip for dl-stream's README (keep in mind that dl-stream is meant as a generic utility, not an add-on of client-zip). And it doesn't say much about dl-stream ; this is just using vanilla JavaScript to adapt to a library, rather than using the library's feature in a certain way.

However, I have now enabled [Discussions on dl-stream](https://github.com/Touffy/dl-stream/discussions) and you're welcome to move/duplicate/point to this over there.

**Reply by spookyuser** [reply]
Ah make sense, thanks again!
